src/dj.erl

```diff
@@ -1,222 +1,758 @@
+        , prop_list/1
+        , to_map/1
+          %% Lists
+        , list/1
+        , index/2
+        , sequence/1
+          %% Manipulating decoders
+        , map/2
+        , chain/2
+        , fail/1
+        , succeed/1
+          %% Fancy decoders
+        , mapn/2
+        , exactly/2
         , one_of/1
-        , list_of/1
-          %% conversion
-        , to_atom/0
-        , to_existing_atom/0
-          %% utility
-        , lift/1
-        ]
-       ).
+          %% Running a decoder
+        , decode/2
+        , decode/3
+        ]).
+
+%%%-----------------------------------------------------------------------------
+%%% Types
+%%%-----------------------------------------------------------------------------
+
+-opaque decoder(T) :: fun ((jsx:json_term()) -> result(T, [E :: error(), ...])).
+
+-type result(V, E) :: {ok, V} | {error, E}.
+
+-type error() :: {unexpected_type, ExpectedType :: type(), jsx:json_term()}
+               | {missing_field, field(), jsx:json_term()}
+               | {missing_index, non_neg_integer(), jsx:json_term()}
+               | {in_field, field(), [error(), ...]}
+               | {at_index, non_neg_integer(), [error(), ...]}
+               | {custom, any()}
+               | {invalid_json, any()}.
+
+-type type() :: binary
+              | integer
+              | pos_integer
+              | neg_integer
+              | non_neg_integer
+              | null
+              | boolean
+              | float
+              | map
+              | list.
+
+-type field() :: atom() | binary().
+
+-export_type([ decoder/1
+             , result/2
+             , error/0
+             , type/0
+             , field/0
+             ]).
 
 %%%-----------------------------------------------------------------------------
 %%% API
 %%%-----------------------------------------------------------------------------
 
-decode(Json, Decoders) ->
-  (compose(Decoders))(Json).
+%% @equiv dj:decode(Json, Decoder, [{labels, existing_atom}])
+-spec decode(Json, decoder(T)) -> result(T, [error(), ...]) when
+    Json :: jsx:json_text().
+decode(Json, Decoder) ->
+  decode(Json, Decoder, [{labels, existing_atom}]).
 
-object() ->
-  object([]).
+%% @doc Run a {@type decoder(T)} against arbirary JSON.
+%%
+%% The resulting {@type result(T, error())} is either a tuple `{ok, T}' or a
+%% tuple `{error, error()}' where {@type error()} represents whatever went wrong
+%% during the decoding/validation/transformation process.
+%%
+%% `Opts` are passed on to `jsx:decode/2'. The option `return_maps' is always
+%% added by {@module} and does not need to be specified manually.
+%%
+%% Use of the functions that create {@type decoder(T)}s and functions that help
+%% with composition are discussed individually.
+-spec decode(Json, decoder(T), Opts) -> result(T, error()) when
+    Json :: jsx:json_text(),
+    Opts :: [term()].
+decode(Json, Decoder, Opts) ->
+  try
+    V = jsx:decode(Json, [return_maps | Opts]),
+    Decoder(V)
```

Comment:
Perhaps this should be moved out of the try/catch so _only_ errors thrown by `jsx` become an `invalid_json` error.